### ISSUE TITLE
how to disable the bp of the bottom layers ?

### ISSUE THREAD

**Issue by starimpact** [apache/incubator-mxnet#8074]
hi,
i know that it can fix parameters of network, but it still do bp on the feature layers. so, does mxnet support to block the bp of bottom layers of network?
for example:
a network has 8 layers, but i just want to train the 8'th layer, and the pre 7 layers bp will be blocked.
how to do it in an easy way?
i know a method which uses two modules...

**Comment by chinakook** [comment]
attr={'lr_mult': '0'}

**Comment by starimpact** [comment]
hi, guy , I searched the 'lr_mult' in the whole mxnet source code. it only appeared in *.py files. so , I don't think this is the answer.

**Comment by chinakook** [comment]
bn = mx.sym.BatchNorm(data=conv, attr={'lr_mult': '0.1'})

**Comment by starimpact** [comment]
it is just a learning rate multiplier, it is not what I am looking for.

**Comment by ZiyueHuang** [comment]
`mx.sym.BlockGrad` makes the outermost grads zeros and be able to bp without effects. `mx.sym.MakeLoss` makes a node be able to bp too but with loss effects, for example, `mx.sym.MakeLoss(A * B)`.

**Comment by starimpact** [comment]
i know BlockGrad can make BP has no effect, but it can not stop BP.

**Comment by ZiyueHuang** [comment]
Then I think you should hack gradient pass in nnvm because this pass ensures that every head bp.

**Comment by starimpact** [comment]
yeah.....

**Comment by Godricly** [comment]
check faster rcnn example in using prefix param. It might work for you.

**Comment by zhreshold** [comment]
The computation graph is a graph, not stacks,  I don't think there's a general enough condition to determine which layer should be freeze or not unless the network has no branch.
Tl;DR, you have to specify layers by their names.

**Comment by starimpact** [comment]
specifying layers name is an option.
but for the most network architectures, they could have many branches, but only one torso. we just need to specify the layer name which is in the torso, then the bottom of the layers behind it will be freezed.

**Comment by starimpact** [comment]
@szha i am not sure whether this function can be solve in the python side.

**Comment by starimpact** [comment]
@Godricly i know prefix parameter exactly, it is not what I need.

**Comment by Godricly** [comment]
Maybe we need  an attribute, like the one in memonger, for each operator to determine if bp is necessary. We should be able to derive if the bp is needed. Also some cases exist that bp is required but param updates are not wanted.

**Comment by thomelane** [comment]
@yzhliu good to close given the suggestions, and option of `autograd.record` scoping given by @ThomasDelteil, thanks! @starimpact please reopen if still an issue.
